Topics/LSP.md

```diff
@@ -0,0 +1,42 @@
+
+# The Liskov Substitution Principle
+​
+## 1. Motivation for The Liskov Substitution Principle (LSP)
+You may know of the LSP from the SOLID principles, which roughly says that instances of a subclass should be able to be substituted for an instance of its superclass without breaking the program. This is a powerful idea; but in fact, the LSP proposes a stronger notion of subtyping:
+
+_Subtype Requirement_: Let ![{\displaystyle \phi (x)}](https://wikimedia.org/api/rest_v1/media/math/render/svg/546b660b2f3cfb5f34be7b3ed8371d54f5c74227) be a property provable about objects ![{\displaystyle x}](https://wikimedia.org/api/rest_v1/media/math/render/svg/87f9e315fd7e2ba406057a97300593c4802b53e4) of type T. Then ![{\displaystyle \phi (y)}](https://wikimedia.org/api/rest_v1/media/math/render/svg/db7ffe2f7daf9bae8d3f2711b2fd67348aceb3dc) should be true for objects ![{\displaystyle y}](https://wikimedia.org/api/rest_v1/media/math/render/svg/b8a6208ec717213d4317e666f1ae872e00620a0d) of type S where S is a subtype of T.
+
+If the subtype requirement is satisfied, then not only is the correctness of our program preserved when substituting objects for sub-objects, but also any *provable* property! What is means for a property to be provable is perhaps unclear, so let's just assume for simplicity our own intuitive notions of provability.
+
+## 2. The Principles in The LSP
+The subtype requirement is an incredibly powerful tool for reasoning about our programs, but how does the LSP help us guide our programming to satsify it? The LSP imposes requirements[^1] on method signatures common across many typed languages:
+
+ 1. **Contravariance of parameter types in the subtype** – if parameters **P** are permissible in method ***f*** of a type, then **P** is also be permissible in ***f*** of any subtype. Equivalently:
+	- If it "works" for a method in a type, then it "works" for the same method in any subtype.
+	- The space of permissible parameters of the subtype is a **superset** of the space of permissible parameters of the type.
+2. **Covariance of return types in the subtype** – a requirement symmetric to the first:
+	- The return type of a method in the subtype should "work" as a return type to the same method in the type.
+	- The space of permissible return values of the subtype is a **subset** of the space of permissible return values of the type.
+3. **New exceptions cannot be thrown by the methods in the subtype, except if they are subtypes of exceptions thrown by the methods of the supertype**
+
+but also imposes additional behavioural requirements[^1]:
+
+4. **Preconditions cannot be strengthened in the subtype.**
+5. **Postconditions cannot be weakened in the subtype.**
+6. **Invariants cannot be weaked in the subtype.**
+7. **History Constraint:** subtypes should not modify state defined in the type in a manner that is not expected by the type.
```

Comment:
I think it would be too much to go over for this article. Doing so would add a lot fo code bloat, which I think would make the overall article much more difficult to read. The paragraph below does go into the History Constraint, but the importance of the other requirements should not be too difficult to understand.